_drafts/2023-06-01-functional-error-handling-in-kotlin-part-2.md

```diff
@@ -0,0 +1,912 @@
+
+```kotlin
+val jobCompany: String = appleJob.map { it.company.name }.getOrElse { "Unknown company" }
+```
+
+The `getOrElse` function takes a lambda with the error as a parameter, so we can use it to react differently to different errors:
+
+```kotlin
+val jobCompany2: String = 
+    appleJob.map { it.company.name }.getOrElse { jobError -> 
+        when (jobError) {
+            is JobNotFound -> "Job not found"
+            is GenericError -> "Generic error"
+        }
+}
+```
+
+Using typed errors has many advantages:
+
+ 1. We can use the type system to check if all the possible cases are handled.
+ 2. The possible causes of failure are listed directly in the function's signature as the left part of the `Either` type. Understanding the possible causes of failure lets us build better tests and error-handling strategies.
+ 3. Typed errors compose better than exceptions.
+
+To prove the above advantages, as we previously did for the `Result` type, it's time to use the `Either` type in our example. Let's change the `Jobs` module to return an `Either` type instead of a `Result` type:
+
+```kotlin
+interface Jobs {
+
+    fun findById(id: JobId): Either<JobError, Job>
+}
+```
+
+Here we are using the `JobError` ADT we defined so far. At this point, we can know precisely how the function can fail by looking at the signature. Now, we can implement the `LiveJobs` class:
+
+```kotlin
+class LiveJobs : Jobs {
+
+    override fun findById(id: JobId): Either<JobError, Job> =
+        try {
+            JOBS_DATABASE[id]?.right() ?: JobNotFound(id).left()
+        } catch (e: Exception) {
+            GenericError(e.message ?: "Unknown error").left()
+        }
+} 
+```
+
+As we might expect, we're wrapping the happy path with a `Right` instance and all the available error cases with a `Left` instance. We are treating the absence of a job as a logic error, and we're wrapping it with a `JobNotFound` object using the recommended and idiomatic syntax:
+
+```kotlin
+value?.right() ?: error.left()
+```
+
+On the other hand, we catch all the exceptions and wrap them in a `GenericError` instance. The pattern of catching exceptions and wrapping them in a `Left` instance is so common that Arrow provides the `catch` function to do it for us in the companion object of the `Either` type:
+
+```kotlin
+@JvmStatic
+@JvmName("tryCatch")
+public inline fun <R> catch(f: () -> R): Either<Throwable, R> =
+  try {
+    f().right()
+  } catch (t: Throwable) {
+    t.nonFatalOrThrow().left()
+  }
+```
+
+The `nonFatalOrThrow` function checks whether the exception should be handled. The fatal exceptions are the following and their subclasses:
+
+* `VirtualMachineError`
+* `ThreadDeath`
+* `InterruptedException`
+* `LinkageError`
+* `ControlThrowable`
+* `CancellationException`
+
+In detail, it's essential not catching `CancellationException` because it's used by Kotlin to [cancel coroutines](https://blog.rockthejvm.com/kotlin-coroutines-101/#7-cancellation).
```

Comment:
This sentence is very useful but slighly unrelated to the article. You can write it as a "by the way"/small digression that ties in to the general error handling topic:

_The subtypes of these errors should not be caught. For example, `CancellationException` should not be caught because it's used by Kotlin to [cancel coroutines](https://blog.rockthejvm.com/kotlin-coroutines-101/#7-cancellation) can catching it can break normal functioning of coroutines._